File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    answer = []
    mat = [[0] * n for _ in range(n)]
    x = 0
    y = 0
    cur = 1
    
    while cur <= (n ** 2 + n) / 2:
        while y < n and mat[y][x] == 0:
            mat[y][x] = cur
            cur += 1
            y += 1
        
        y -= 1
        x += 1
        while x < n and mat[y][x] == 0:
            mat[y][x] = cur
            cur += 1
            x += 1
            
        x -= 2
        y -= 1
        while x >= 0 and y >= 0 and mat[y][x] == 0:
            mat[y][x] = cur
            cur += 1
            x -= 1
            y -= 1
        
        x += 1
        y += 2

    for idx, arr in enumerate(mat):
        answer.extend(arr[:idx + 1])
        
    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    const mat = [...Array(n)].map(_ => [...Array(n)].map(_ => 0));
    let [y, x, cur] = [-1, 0, 1];
    
    for (let i = 0; i < n; i++) {   // 직선 경로의 개수는 총 n개
        for (let k = 0; k < n - i; k++) {   // i번째 직선 경로의 길이는 (n - i)
            if (i % 3 === 0) {
                y += 1;
            }
            
            if (i % 3 === 1) {
                x += 1;
            }
            
            if (i % 3 === 2) {
                y -= 1;
                x -= 1;
            }
            
            mat[y][x] = cur;
            cur += 1;
        }
    }
    
    return mat.map((arr, idx) => arr.slice(0, idx + 1)).reduce((acc, cur) => [...acc, ...cur]);
}